### PR TITLE
feat(sdk): backport PipelineTaskFinalStatus change to `sdk/release-1.8` branch

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
Re-implement https://github.com/kubeflow/pipelines/pull/7309 in `sdk/release-1.8` branch.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by chensun** [comment]
> @chensun: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> kubeflow-pipelines-samples-v2	[8ccb372](https://github.com/kubeflow/pipelines/commit/8ccb372b99b8ee55d4ccb52a078ea9a83005b459)	[link](https://oss-prow.knative.dev/view/gs/oss-prow/pr-logs/pull/kubeflow_pipelines/7483/kubeflow-pipelines-samples-v2/1509061066556968960)	true	`/test kubeflow-pipelines-samples-v2`

Again, going to ignore this test failure.

File: sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_task_final_status.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+from kfp.v2 import dsl
+from kfp.v2 import compiler
+from kfp.v2.dsl import component
+
+from kfp.v2.dsl import PipelineTaskFinalStatus
+
+
+@component
+def exit_op(user_input: str, status: PipelineTaskFinalStatus):
```
Comment (connor-mccarthy) [inline sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_task_final_status.py:24]: nit: I think this should be `dsl.PipelineTaskFinalStatus` per [Google Python Style Guide 2.2 Imports](https://google.github.io/styleguide/pyguide.html#22-imports).

I haven't been entirely consistent about this in my own PRs, FWIW.

**Review by connor-mccarthy — COMMENTED** [review]
/lgtm

This looks great, @chensun. Just some small nitpick comments and things to consider.

File: sdk/python/kfp/v2/components/executor.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 from typing import Any, Callable, Dict, List, Optional, Union
 
 from kfp.v2.components.types import artifact_types, type_annotations
+from kfp.v2.components import task_final_status
```
[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/v2/components/executor.py:19]: Thank you for explaining -- that makes sense to me.

File: sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_task_final_status.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_task_final_status.py:24]: I asked the same question before. :) 
The answer was it's okay to break the style guide to keep user experience simple, and I think that makes sense.
For the same reason, you may see our public examples like below.
```python
from kfp.dsl import Input, Dataset

@component
def my_component(data: Input[Dataset]):
   ...
```

File: sdk/python/kfp/v2/components/task_final_status.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import dataclasses
+from typing import Optional
+
+
+@dataclasses.dataclass
+class PipelineTaskFinalStatus:
+    """The final status of a pipeline task.
+
+    This is the Python representation of the proto: PipelineTaskFinalStatus
+    https://github.com/kubeflow/pipelines/blob/61d9210c61ff9be780b678ab78a7f7200d50cffa/api/v2alpha1/pipeline_spec.proto#L886
```
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/components/task_final_status.py:25]: Good catch. And I'll update the master branch on this shortly.

File: sdk/python/kfp/components/_components.py [hunk 4]
```diff
@@ -30,6 +30,8 @@
 from .structures import *
 from ._data_passing import serialize_value, get_canonical_type_for_type_name
 
+from kfp.v2.components.types import type_utils
```
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/components/_components.py:33]: Adding current year is optional per https://opensource.google/documentation/reference/copyright
I'll skip that for this branch. :)

**Pull request merged**